libcxx/include/__type_traits/remove_pointer.h

```diff
@@ -17,7 +17,8 @@
 
 _LIBCPP_BEGIN_NAMESPACE_STD
 
-#if !defined(_LIBCPP_WORKAROUND_OBJCXX_COMPILER_INTRINSICS) && __has_builtin(__remove_pointer)
+#if !defined(_LIBCPP_WORKAROUND_OBJCXX_COMPILER_INTRINSICS) && __has_builtin(__remove_pointer) &&                      \
+    defined(_LIBCPP_COMPILER_CLANG_BASED)
```

Comment:
`_LIBCPP_COMPILER_CLANG_BASED` seems to be quite heavy restriction. It's better to just restrict GCC 14.1 and add a comment why GCC 14 is disabled. Once GCC 15.1 is released next year we can reevaluate its status.

FYI GCC 14 is not yet available in [Ubuntu PPA](https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test/+packages?field.name_filter=gcc&field.status_filter=published&field.series_filter=). Once it's available we can start testing GCC 14.1 is our CI.

Reply:
I disagress with that. Along with this, GCC has tons of new builtins in trunk targeting for 15.1 (or 14.2). They might be breaking things as well. There is no guarentee that other compilers other than clang will follow clang's way of handling builtins and will make changes finally. For example, below code doesn't changes since 2013.

```
#if __cplusplus > 201103L
  /// Alias template for remove_pointer
  template<typename _Tp>
    using remove_pointer_t = typename remove_pointer<_Tp>::type;

  /// Alias template for add_pointer
  template<typename _Tp>
    using add_pointer_t = typename add_pointer<_Tp>::type;
#endif
```

https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/std/type_traits#L2146

For the same time, gcc 14.1 is available in archlinux and made as default compiler. That's what push me to make this change.

Reply:
![image](https://github.com/llvm/llvm-project/assets/54673341/27a6afca-bbaf-4c01-99bd-b12c2e313c55)